### PR TITLE
Autoload with share button and hotkeys

### PR DESCRIPTION
**Pull request by fairnightzz**

### CODE REVIEW

File: src/components/MangaControl.tsx [hunk 1]
```diff
@@ -55,8 +61,13 @@ const MangaControl: React.FunctionComponent<Props> = (props: Props) => {
   const changePage = (event: React.ChangeEvent<HTMLSelectElement>) => {
     updatePage(+event.currentTarget.value);
   };
+  const changeAutoLoadChapter = () => {
```
Comment (MrPicklePinosaur) [inline src/components/MangaControl.tsx]: also should rename to `toggleAutoLoadChapter`

File: src/components/MangaControl.tsx [hunk 2]
```diff
@@ -55,8 +61,13 @@ const MangaControl: React.FunctionComponent<Props> = (props: Props) => {
   const changePage = (event: React.ChangeEvent<HTMLSelectElement>) => {
     updatePage(+event.currentTarget.value);
   };
+  const changeAutoLoadChapter = () => {
+    updateReaderSettings({ ...readerSettings, autoLoadChapter: !readerSettings.autoLoadChapter })
```
Comment (MrPicklePinosaur) [inline src/components/MangaControl.tsx]: seem to be using this function in the form
```
updateReaderSettings({...oldReaderSettings, autoLoadChapter: blah})
```
a lot.

could potentially refactor `updateReaderSettings` to take in a partial reader setting object and destructure it like so:
```
const updateReaderSettings = (newSettings) => (
   newReaderSettings = {
     ...oldReaderSettings,
     ...newSettings
  }
)
```
then you can just use it like
```
updateReaderSettings({autoLoadChapter: blah})
```

File: src/components/Modal.tsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                {title}
+              </Dialog.Title>
+              <div className="mt-2">
+                {children}
+              </div>
+
+              <div className="mt-4">
+                <button
+                  type="button"
+                  className="inline-flex justify-center px-4 py-2 text-sm font-medium text-blue-900 bg-blue-100 border border-transparent rounded-md hover:bg-blue-200 focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 focus-visible:ring-blue-500"
```
Comment (MrPicklePinosaur) [inline src/components/Modal.tsx:53]: maybe start using tailwind @ apply ?

File: src/pages/Reader.tsx [hunk 4]
```diff
@@ -56,11 +58,40 @@ class Reader extends React.Component<any, State> {
...
       catch (err) {
         console.log('No settings found.')
       }
     }
+    const keyBindings = {
+      previousPage: "K",
+      nextPage: "J",
+      previousChapter: "H",
+      nextChapter: "L",
+    }
```
Comment (MrPicklePinosaur) [inline src/pages/Reader.tsx:61]: why shift+hjkl?
also for full page navigation `d` and `u` are more commonly used. `j` and `k` are just for small nudges.

and then for page nav i think `n` and `p` might be better?

Comment (MrPicklePinosaur) [reply src/pages/Reader.tsx:61]: hjkl are fine too but i think it shouldn't be shift+hjkl

File: src/pages/Reader.tsx [hunk 5]
```diff
@@ -208,19 +251,47 @@ class Reader extends React.Component<any, State> {
...
+  setKeyBindings = (newKeyBindings: KeyBindings) => {
+    this.setState(curState => ({
+      ...curState,
+      keyBindings: newKeyBindings
+    }));
+    localStorage.setItem("carrotKeyBindings", JSON.stringify(newKeyBindings))
+  }
+
+  handleHotKeys = (event: KeyboardEvent) => {
+    const key = event.key.toUpperCase();
```
Comment (MrPicklePinosaur) [inline src/pages/Reader.tsx:244]: why `toUpperCase`

File: src/types/reader.types.ts [hunk 6]
```diff
@@ -25,9 +26,19 @@ export enum Menu {
   open = "open",
   close = "close"
 }
+
+export type AutoLoadChapter = boolean;
```
Comment (MrPicklePinosaur) [inline src/types/reader.types.ts]: why did you need to define a type for this

File: src/pages/Reader.tsx [hunk 7]
```diff
@@ -56,11 +58,40 @@ class Reader extends React.Component<any, State> {
...
+    const keyBindings = {
+      previousPage: "K",
+      nextPage: "J",
+      previousChapter: "H",
+      nextChapter: "L",
+    }
+    const keyBindingsLocal = localStorage.getItem('carrotKeyBindings')
+    if (keyBindingsLocal) {
+      try {
+        const keyBindingsParsed = JSON.parse(keyBindingsLocal);
```
Comment (MrPicklePinosaur) [inline src/pages/Reader.tsx]: actually i still dont like how functions like these are handled.

a possible alternative is to define sm like
```
type Nullable<T> = { [K in keyof T]: T[K] | null };
```
and then do
```
const keyBindingsParsed: Nullable<Keybindings> = JSON.parse(keyBindingsLocal);
```
so now this makes sure that the localstorage object contains only the keys that are present in Keybindings. and then you can just object destructure
```
keybindings = {
  ...keybindingsDefault,
  ...keybindingsParsed
}
```

**Review by MrPicklePinosaur — CHANGES_REQUESTED** [review]
review comments

File: src/pages/Reader.tsx (hunk 4 above)
Comment (fairnightzz) [reply src/pages/Reader.tsx:61]: Where are u getting shift from

File: src/components/Modal.tsx (hunk 3 above)
Comment (fairnightzz) [reply src/components/Modal.tsx:53]: :(

File: src/pages/Reader.tsx (hunk 5 above)
Comment (fairnightzz) [reply src/pages/Reader.tsx:244]: Oh no wonder if think it's uppercase because they're capitals. I meant for them to just be the key, no need for shift. Should they be displayed as lowercase instead

File: src/types/reader.types.ts (hunk 6 above)
Comment (fairnightzz) [reply src/types/reader.types.ts]: It's part of the settings definition and just makes sense to define it like the others

File: src/pages/Reader.tsx (hunk 7 above)
Comment (fairnightzz) [reply src/pages/Reader.tsx]: I think I get what u mean, the nullable part is a bit confusing tho

File: src/components/MangaControl.tsx (hunk 2 above)
Comment (fairnightzz) [reply src/components/MangaControl.tsx]: wait what type is it going to be tho, not too sure how to define newSettings type

Comment (MrPicklePinosaur) [reply src/components/MangaControl.tsx]: probably `Nullable<SettingsType>`

File: src/pages/Reader.tsx (hunk 4 above)
Comment (MrPicklePinosaur) [reply src/pages/Reader.tsx:61]: since its capital HJKL

File: src/pages/Reader.tsx (hunk 5 above)
Comment (MrPicklePinosaur) [reply src/pages/Reader.tsx:244]: yeah display them as lowercase

File: src/types/reader.types.ts (hunk 6 above)
Comment (MrPicklePinosaur) [reply src/types/reader.types.ts]: i think thats overdoing it. it makes sense for the other ones since they are enums, but i think
```
autoLoadChapter: boolean
```
is way more clear than
```
autoLoadChapter: AutoLoadChapter
```

Comment (fairnightzz) [reply src/types/reader.types.ts]: Sure

File: src/pages/Reader.tsx (hunk 7 above)
Comment (MrPicklePinosaur) [reply src/pages/Reader.tsx]: its defining a type where each property can be nullable, ex:
original type:
```
type MyType = {
  propOne: string,
  propTwo: boolean,
  propThree: number
}
```
then `Nullable<MyType>` will look like
```
type MyType = {
  propOne: string | null,
  propTwo: boolean | null,
  propThree: number | null
}
```

Comment (fairnightzz) [reply src/pages/Reader.tsx]: Is null considered as undefined

Comment (MrPicklePinosaur) [reply src/pages/Reader.tsx]: ah my bad, this should be undefined,
so
```
type Partial<T> = { [K in keyof T]: T[K] | undefined };
```
and naming it partial instead might make more sense, since it makes a version of the `T` object but each key is optional

Comment (fairnightzz) [reply src/pages/Reader.tsx]: K

File: src/components/MangaControl.tsx [hunk 8]
```diff
@@ -35,7 +41,7 @@ type Props = {
 }
 
 const MangaControl: React.FunctionComponent<Props> = (props: Props) => {
-  const { readerSettings, updateReaderSettings, chapter, chapters, updateChapter, page, pages, updatePage } = props;
+  const { readerSettings, updateReaderSettings, keyBindings, setKeyBindings, chapter, chapters, updateChapter, page, pages, updatePage } = props;
   const toggleMenuState = () => {
     updateReaderSettings({ ...readerSettings, menu: readerSettings.menu === Menu.open ? Menu.close : Menu.open, });
```
Comment (MrPicklePinosaur) [inline src/components/MangaControl.tsx]: also is readerSettings.menu ever going to have more states than just open or close?

if not you could just change it to menuOpen or menuIsOpen and have it be a bool

File: src/components/Modal.tsx (hunk 3 above)
Comment (MrPicklePinosaur) [reply src/components/Modal.tsx:53]: u can do a seperate PR to move a bunch of long css into their own classes

**Comment by MrPicklePinosaur** [comment]
ready for merge, find out if you can delete the branch when u merge
